quinn-proto/src/connection/mod.rs

```diff
@@ -760,10 +784,104 @@ where
                     self.path.challenge_pending = false;
                 }
                 Timer::Pacing => trace!("pacing timer expired"),
+                Timer::PushNewCid => {
+                    let num_pushed_cids = 1;
+                    let mut ids_found = false;
+                    for id in self.retired_prior_to_seq..self.retire_cid_seq {
+                        ids_found |= self.cids_active_seq.contains(&id);
+                    }
+                    if !ids_found {
+                        self.retired_prior_to_seq = self.retire_cid_seq;
+                    }
+
+                    let next_checkpoint = self.cids_timeline.pop_front();
+                    let next_seq = next_checkpoint.unwrap().sequence + 1;
+                    let current_seq = self.retire_cid_seq;
+                    //  Endpoints SHOULD NOT issue updates of the Retire Prior To field
+                    //  before receiving RETIRE_CONNECTION_ID frames that retire all
+                    //  connection IDs indicated by the previous Retire Prior To value.
+                    // https://tools.ietf.org/html/draft-ietf-quic-transport-29#section-5.1.2
+                    if next_seq > self.retire_cid_seq && !ids_found {
+                        self.retire_cid_seq = next_seq;
+                    }
+                    // An endpoint MUST NOT
+                    // provide more connection IDs than the peer's limit.
+                    //
+                    // An endpoint MAY
+                    // send connection IDs that temporarily exceed a peer's limit if the
+                    // NEW_CONNECTION_ID frame also requires the retirement of any excess,
+                    // by including a sufficiently large value in the Retire Prior To field.
+                    // https://tools.ietf.org/html/draft-ietf-quic-transport-29#section-5.1.1
+                    // In this case, `retire prior to` must increase to accommodate new cid
+                    let must_increase_retire_cid =
+                        self.cids_active_seq.len() as u64 >= self.peer_params.issue_cids_limit();
+                    if must_increase_retire_cid
+                        && self.retire_cid_seq - current_seq < num_pushed_cids
```

Comment:
What's the second clause here doing?

Reply:
```
let must_increase_retire_cid = self.cids_active_seq.len() as u64 >= self.peer_params.issue_cids_limit();
```
Check if the current number of active cid is already the maximum set by peer. It is possible that PushNewCID timer fires but we haven't received any RETIRE_CONNECTION_ID frame from peer. 

There are two statements in the RFC: 1. Endpoints SHOULD NOT issue updates of the Retire Prior To field before receiving RETIRE_CONNECTION_ID frames that retire all connection IDs indicated by the previous Retire Prior To value. 2. An endpoint MUST NOT provide more connection IDs than the peer's limit.

In the above case, we have to retire more cids since the `PushNewCID` timer fires which causes some contradiction between  the statement 1 and 2. From one side, we cannot advance `retire_prior_to` because of (1) but meanwhile we have to advance this field because of (2) since all unretired cids are considered active. Since MUST overrides SHOULD, and there is another RFC statement `An endpoint MAY send connection IDs that temporarily exceed a peer's limit if the NEW_CONNECTION_ID frame also requires the retirement of any excess`, we use `must_increase_retire_cid ` to indicate whether we end up with this awkward situation and have to increase the value in the field without other choice.

Reply:
```
 if must_increase_retire_cid && self.retire_cid_seq - current_seq < num_pushed_cids
```

This is a safety check which makes sure the difference between last and current `retire_perior_to` is more than `num_pushed_cids `. It means we won't issue more cids than retired.

Reply:
I have to admit the naming is messy and somehow awful. Working on a new version with better readability.

Reply:
I think the processing logic is too complicated for a realistic scenario. Will simplify in the next iteration